fv3gfs/util/halo_updater.py

```diff
@@ -0,0 +1,263 @@
+    from .communicator import Communicator
+
+
+class HaloUpdater:
+    """Exchange halo information between ranks.
+
+    The class is responsible for the entire exchange and uses the __init__
+    to precompute the maximum of information to have minimum overhead at runtime.
+    Therefore it should be cached for early and re-used at runtime.
+
+    - from_scalar_specifications/from_vector_specifications are used to create an HaloUpdater
+      from a list of memory specifications
+    - update and start/wait trigger the halo exchange
+    - the class creates a "pattern" of exchange that can fit any memory given to do/start
+    - temporary references to the Quanitites are held between start and wait
+    - Optional: finalize_on_wait allows for a behavior that flushes buffers at every wait call
+        and re-fetch buffers at every start. This allows to keep the API behavior consistent with
+        previous version when using Communicator.start_halo_update/wait
+    """
+
+    def __init__(
+        self,
+        comm: "Communicator",
+        tag: int,
+        transformers: Dict[int, HaloDataTransformer],
+        timer: Timer,
+        finalize_on_wait: bool,
+    ):
+        self._comm = comm
+        self._tag = tag
+        self._transformers = transformers
+        self._timer = timer
+        self._recv_requests: List[AsyncRequest] = []
+        self._send_requests: List[AsyncRequest] = []
+        self._inflight_x_quantities: Optional[Tuple[Quantity, ...]] = None
+        self._inflight_y_quantities: Optional[Tuple[Quantity, ...]] = None
+        self._finalize_on_wait = finalize_on_wait
+
+    def __del__(self):
+        """Clean up all buffers on garbage collection"""
+        if (
+            self._inflight_x_quantities is not None
+            or self._inflight_y_quantities is not None
+        ):
+            raise RuntimeError(
+                "An halo exchange wasn't completed and a wait() call was expected"
+            )
+        if not self._finalize_on_wait:
+            for buffer in self._transformers.values():
+                buffer.finalize()
+
+    @classmethod
+    def from_scalar_specifications(
+        cls,
+        comm: "Communicator",
+        numpy_like_module: NumpyModule,
+        specifications: Iterable[HaloUpdateSpec],
+        boundaries: Iterable[Boundary],
+        tag: int,
+        optional_timer: Optional[Timer] = None,
+        finalize_on_wait: bool = False,
+    ) -> "HaloUpdater":
+        """Create/retrieve as many packed buffer as needed and queue the slices to exchange.
+
+        Args:
+            comm: communicator to post network messages
+            numpy_like_module: module implementing numpy API
+            specifications: data specifications to exchange, including number of halo points
+            boundaries: informations on the exchange boundaries.
+            tag: network tag (to differentiate messaging) for this node.
+            optional_timer: timing of operations.
```

Comment:
Please add `finalize_on_wait` to docstring.

Reply:
Removed.